### ISSUE TITLE
Implement: Align Your Steps: Optimizing Sampling Schedules in Diffusion Models

### ISSUE THREAD

**Issue by joe-aivatarz** [huggingface/diffusers#7760]
Align Your Steps: Optimizing Sampling Schedules in Diffusion Models is a general and principled approach to optimizing the sampling schedules of DMs for high-quality outputs.
This work is presented by Nvidia labs in this paper:
https://arxiv.org/abs/2404.14507

and the project page is here
https://research.nvidia.com/labs/toronto-ai/AlignYourSteps/

the page propose a very small change that have big impact on the inference quality

**Comment by DN6** [comment]
Hmm the results do look good. But from the project page they mention 

> We leverage methods from stochastic calculus and find optimal schedules specific to different solvers, trained DMs and datasets.

So each scheduler would have it's own set of aligned steps. We can support this by allowing setting the timesteps directly for all schedulers. 

@yiyixuxu wdyt? The Colab notebook example the authors have provided uses diffusers and modifies the DPM scheduler 

```python
from diffusers import DPMSolverMultistepScheduler as DefaultDPMSolver

# Add support for setting custom timesteps
class DPMSolverMultistepScheduler(DefaultDPMSolver):
    def set_timesteps(
        self, num_inference_steps=None, device=None,
        timesteps=None
    ):
        if timesteps is None:
            super().set_timesteps(num_inference_steps, device)
            return

        all_sigmas = np.array(((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5)
        self.sigmas = torch.from_numpy(all_sigmas[timesteps])
        self.timesteps = torch.tensor(timesteps[:-1]).to(device=device, dtype=torch.int64) # Ignore the last 0

        self.num_inference_steps = len(timesteps)

        self.model_outputs = [
            None,
        ] * self.config.solver_order
        self.lower_order_nums = 0

        # add an index counter for schedulers that allow duplicated timesteps
        self._step_index = None
        self._begin_index = None
        self.sigmas = self.sigmas.to("cpu")  # to avoid too much CPU/GPU communication
```
 
Colab: 
https://colab.research.google.com/drive/1cIwbbO4HRP1aUQ8WcbQBaT8p3868k7BC?usp=sharing#scrollTo=X8WJXsHUV96k

**Comment by chuck-ma** [comment]
I'm wondering how to modify EDMDPMSolverMultistepScheduler to get similar result.
I write a script, but it doen't work. Wondering how to fix it

`
import numpy as np


class EDMDPMSolverMultistepScheduler(DefaultDPMSolver):
    def set_timesteps(
        self, num_inference_steps: int = None, device: torch.device = None,
        timesteps: torch.Tensor = None  
    ):
        if timesteps is None:
            super().set_timesteps(num_inference_steps, device)
            return
        self.num_inference_steps = len(timesteps)
        ramp = np.linspace(0, 1, self.num_inference_steps)
        sigmas = self._compute_sigmas(ramp)


        sigmas = torch.from_numpy(sigmas).to(dtype=torch.float32, device=device)

        if self.config.final_sigmas_type == "sigma_min":
            sigma_last = self.config.sigma_min
        elif self.config.final_sigmas_type == "zero":
            sigma_last = 0
        else:
            raise ValueError(
                f"`final_sigmas_type` must be one of 'zero', or 'sigma_min', but got {self.config.final_sigmas_type}"
            )

        self.sigmas = torch.cat([sigmas, torch.tensor([sigma_last], dtype=torch.float32, device=device)])

        # print("device=", device, "timesteps=", timesteps[:-1])
        self.timesteps = torch.tensor(timesteps[:-1]).to(device=device, dtype=torch.int64)
        # self.sigmas = selected_sigmas.to(device)

        self.num_inference_steps = len(timesteps)


        self.model_outputs = [
            None,
        ] * self.config.solver_order
        self.lower_order_nums = 0

        # add an index counter for schedulers that allow duplicated timesteps
        self._step_index = None
        self._begin_index = None
        self.sigmas = self.sigmas.to("cpu")  # to avoid too much CPU/GPU communication`

**Comment by christopher-beckham** [comment]
Having a `set_timesteps` for all relevant schedulers would also make it much easier to implement things like this: https://github.com/huggingface/diffusers/issues/7651

**Comment by yiyixuxu** [comment]
@DN6 

I'm a little bit confused, it says each scheduler would have its own optimized schedule but this is what they provide. Can these timesteps be used for all schedulers for these models? 

cc @asomoza here maybe you have better ideas since it is popular in community

<table>
<thead>
<tr>
<th style="text-align:left">Model</th>
<th style="text-align:center">Schedule (noise levels)</th>
<th style="text-align:center">Schedule (timestep indices)</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Stable Diffusion 1.5</td>
<td style="text-align:center">[14.615, 6.475, 3.861, 2.697, 1.886, 1.396, 0.963, 0.652, 0.399, 0.152, 0.029]</td>
<td style="text-align:center">[999, 850, 736, 645, 545, 455, 343, 233, 124,  24,   0]</td>
</tr>
<tr>
<td style="text-align:left">SDXL</td>
<td style="text-align:center">[14.615, 6.315, 3.771, 2.181, 1.342, 0.862, 0.555, 0.380, 0.234, 0.113, 0.029]</td>
<td style="text-align:center">[999, 845, 730, 587, 443, 310, 193, 116,  53,  13,   0]</td>
</tr>
<tr>
<td style="text-align:left">DeepFloyd-IF / Stage-1</td>
<td style="text-align:center">[160.41, 8.081, 3.315, 1.885, 1.207, 0.785, 0.553, 0.293, 0.186, 0.030, 0.006]</td>
<td style="text-align:center">[995, 920, 811, 686, 555, 418, 315, 174, 109,  12,   0]</td>
</tr>
<tr>
<td style="text-align:left">Stable Video Diffusion</td>
<td style="text-align:center">[700.00, 54.5, 15.886, 7.977, 4.248, 1.789, 0.981, 0.403, 0.173, 0.034, 0.002]</td>
<td style="text-align:center">NA</td>
</tr>
</tbody>
</table>




To support this more natively, I think we can extend the `timesteps` argument to `set_timesteps` method to more schedulers; also extend it to the SVD pipeline they mentioned 

I would first figure out what schedulers/pipelines can these optimized steps be applied to and only support these selected schedulers for now.

**Comment by asomoza** [comment]
Initially I thought it only worked for DPM schedulers but I've been testing it in comfyui and they enabled it for all of them, so far it works with all of them but I think that in all the SDE variants it's a lot worse (not usable).

Also in SDXL it's less noticeable but still get the performance gain.

prompt = "anthropomorphic capybara wearing a suit and working with a computer"
ays = 10 steps 
normal = 25 steps

||ays|normal|
|---|---|---|
|Euler|![comfyui_ays_00001_](https://github.com/huggingface/diffusers/assets/5442875/c0af2d62-7ead-44d0-ba43-f673477bfc73)|![comfyui_normal_00001_](https://github.com/huggingface/diffusers/assets/5442875/69909990-698d-43ae-9179-a8f4d47ee223)|
|Heun|![comfyui_ays_00002_](https://github.com/huggingface/diffusers/assets/5442875/1cdb2fee-f9fe-4b49-bb7b-d96d59e0ddc4)|![comfyui_normal_00002_](https://github.com/huggingface/diffusers/assets/5442875/33d8163a-11d9-4860-af96-15621e8eb0a5)|
|dpm_2|![comfyui_ays_00001_](https://github.com/huggingface/diffusers/assets/5442875/8d578e0f-7134-4dfd-b6a2-234f91975a0f)|![comfyui_normal_00001_](https://github.com/huggingface/diffusers/assets/5442875/8f12b702-829b-4000-b0ef-51a4db2c989f)|
|dpmpp_2m|![comfyui_ays_00003_](https://github.com/huggingface/diffusers/assets/5442875/5683cc30-522d-4c7e-bcfe-3cf776f930c9)|![comfyui_normal_00003_](https://github.com/huggingface/diffusers/assets/5442875/0e0b479b-eb21-4fd8-b5e4-3a44e7647286)|
|dpmpp_2m_sde_gpu|![comfyui_ays_00001_](https://github.com/huggingface/diffusers/assets/5442875/c2125724-4ae9-4fca-856f-d60cf16ebc67)|![comfyui_normal_00001_](https://github.com/huggingface/diffusers/assets/5442875/b47d7135-67fd-4dda-8c5f-abeebda037a9)

**Comment by haofanwang** [comment]
I'm curious about whether it works on fewer steps (2 steps, 4 steps, like other distillation methods)? How can we derive the optimized timesteps from scratch?

**Comment by yiyixuxu** [comment]
@haofanwang they described their method in the paper - not sure how easy it is to reproduce & how much compute it would cost

**Comment by DeFek1** [comment]
Is there away to inference more steps than 10?

**Comment by wonkyoc** [comment]
@DeFek1 you can use a linear interpolation in between those timestep indices

**Comment by DeFek1** [comment]
@wonkyoc how? Is there any code i can try?

**Comment by wonkyoc** [comment]
The original repo includes the code or you could use numpy or other linear algebra library.

https://research.nvidia.com/labs/toronto-ai/AlignYourSteps/howto.html

**Comment by a-r-r-o-w** [comment]
Marking as resolved because AYS is now supported, but if I'm mistaken about the status of this issue, please do re-open (cc @asomoza)
